src/lib/reasoners/rel_utils.ml

```diff
@@ -197,182 +197,210 @@ module type Constraint = sig
+      active. *)
+
+  val subst : ex:Explanation.t -> X.r -> X.r -> t -> t
+  (** [subst ~ex p v t] replaces all instances of [p] with [v] in the
+      constraints.
 
       The explanation [ex] justifies the equality [p = v]. *)
 
-  val add : t -> Constraint.t -> t
-  (** [add c cs] adds the constraint [c] to [cs]. *)
+  val notify : X.r -> t -> t
+  (** [notify r t] marks all constraints involving [r] (i.e. all constraints
+      that have [r] as one of their arguments) as pending.
 
-  val propagate_fresh :  t -> Constraint.domain -> t * Constraint.domain
-  (** [propagate_fresh cs acc] propagates the fresh constraints and returns the
-      new domain, as well as a copy of the constraint set with no fresh
-      constraints.
+      This function should be used when the domain of [r] is updated, if
+      domains are tracked for all representatives. *)
+
+  val notify_leaf : X.r -> t -> t
+  (** [notify_leaf r t] marks all constraints that have [r] as a leaf (i.e.
+      all constraints that have at least one argument [a] such that [r] is in
+      [X.leaves a]) as pending.
 
-      Fresh constraints are constraints that were never propagated yet. *)
+      This function should be used when the domain of [r] is updated, if
+      domains are tracked for leaves only. *)
 
-  val fold_r : (X.r -> 'a -> 'a) -> t -> 'a -> 'a
-  (** [fold_r f cs acc] folds [f] over any representative [r] that is currently
-      associated with a constraint (i.e. at least one constraint currently
-      applies to [r]). *)
+  val fold_args : (X.r -> 'a -> 'a) -> t -> 'a -> 'a
+  (** [fold_args f t acc] folds [f] over all the term representatives that are
+      arguments of at least one constraint. *)
+
+  val next_pending : t -> Constraint.t explained * t
+  (** [next_pending t] returns a pair [c, t'] where [c] was pending in [t] and
+      [t'] is identical to [t], except that [c] is no longer a pending
+      constraint.
 
-  val propagate : t -> X.r -> Constraint.domain -> Constraint.domain
-  (** [propagate cs r dom] propagates the constraints associated with [r] in the
-      constraint set [cs] and returns the new domain map after propagation. *)
+      @raise Not_found if there are no pending constraints. *)
 end = struct
-  module IM = Util.MI
   module MX = Shostak.MXH
 
-  module CS = Set.Make(Constraint)
+  module CS = Set.Make(struct
+      type t = Constraint.t explained
+
+      let compare a b = Constraint.compare a.value b.value
+    end)
 
   type t = {
-    cs_set : CS.t ;
-    (*** All the constraints currently active *)
-    cs_map : CS.t MX.t ;
-    (*** Mapping from semantic values to the constraints that involves them *)
-    fresh : CS.t ;
-    (*** Fresh constraints that have never been propagated *)
+    args_map : CS.t MX.t ;
+    (** Mapping from semantic values to constraints involving them *)
+
+    leaves_map : CS.t MX.t ;
+    (** Mapping from semantic values to constraints they are a leaf of *)
+
+    active : CS.t ;
```

Comment:
I guess `active` means  constraint already propagated with its explanation to `CC(X)`? `propagated` may be a better name?

Reply:
`propagated` would be confusing because this set can contain constraints that have actually never been propagated.

`active` means constraints that are, well, active (the CP community would say "posted" but I don't like that term) — they must be true in the model, and will be propagated when the domain of their arguments change.